compiler-core/src/error.rs

```diff
@@ -2450,300 +2001,762 @@ value, or use `let assert` rather than `let`."
+                module,
+                first,
+                second,
+            } => {
+                let text = format!(
+                    "The module `{module}` is defined multiple times.
+
+First:  {first}
+Second: {second}"
+                );
+
+                Diagnostic {
+                    title: "Duplicate module".into(),
+                    text,
+                    hint: None,
+                    level: Level::Error,
+                    location: None,
+                }
+            }
+
+            Error::DuplicateSourceFile { file } => Diagnostic {
+                title: "Duplicate Source file".into(),
+                text: format!("The file `{file}` is defined multiple times."),
+                hint: None,
+                level: Level::Error,
+                location: None,
             },
 
+            Error::FileIo {
+                kind,
+                action,
+                path,
+                err,
+            } => {
+                let err = match err {
+                    Some(e) => {
+                        format!("\nThe error message from the file IO library was:\n\n    {e}\n")
+                    }
+                    None => "".into(),
+                };
+                let text = format!(
+                    "An error occurred while trying to {} this {}:
+
+    {}
+{}",
+                    action.text(),
+                    kind.text(),
+                    path,
+                    err,
+                );
+                Diagnostic {
+                    title: "File IO failure".into(),
+                    text,
+                    hint: None,
+                    level: Level::Error,
+                    location: None,
+                }
+            }
+
+            Error::NonUtf8Path { path } => {
+                let text = format!(
+                    "Encountered a non UTF-8 path '{}', but only UTF-8 paths are supported.",
+                    path.to_string_lossy()
+                );
+                Diagnostic {
+                    title: "Non UTF-8 Path Encountered".into(),
+                    text,
+                    level: Level::Error,
+                    location: None,
+                    hint: None,
+                }
+            }
+
+            Error::GitInitialization { error } => {
+                let text = format!(
+                    "An error occurred while trying make a git repository for this project:
+
+    {error}"
+                );
+                Diagnostic {
+                    title: "Failed to initialize git repository".into(),
+                    text,
+                    hint: None,
+                    level: Level::Error,
+                    location: None,
+                }
+            }
+
+            Error::Type { path, src, error } => self.type_to_diagnostic(
+                path,
+                src,
+                error.first().expect("there should be at least one error"),
```

Comment:
Let's remove this runtime check and instead use the type system to enforce that it is correct. We could use a Vec1 here and have this function return Vec1<Diagnostic>. This would also simplify it as we won't have the two paths through this code with the first special case for Error::Type

Reply:
Switching the error within Error::Type to a vec1 makes sense to me. The main reason I had `to_diagnostic` as a separate thing was to avoid wrapping all the non type cases in vec! (mainly to make the initial refactoring lift easier).

Reply:
Ok between doing this and switching the formatting/structuring to be back to the original the errors are now much closer to the original